src/native/libs/System.Native/pal_runtimeinformation.c

```diff
@@ -48,8 +53,89 @@ int32_t SystemNative_GetUnixVersion(char* version, int* capacity)
     return 0;
 }
 
-/* Returns an int representing the OS Architecture. -1 if same as process architecture. */
+// Keep in sync with System.Runtime.InteropServices.Architecture enum
+enum
+{
+    ARCH_X86,
+    ARCH_X64,
+    ARCH_ARM,
+    ARCH_ARM64,
+    ARCH_WASM,
+    ARCH_S390X,
+    ARCH_LOONGARCH64,
+    ARCH_ARMV6,
+};
+
 int32_t SystemNative_GetOSArchitecture()
 {
-    return -1;
+#ifdef TARGET_WASM
+    return ARCH_WASM;
+#else
+    int32_t result = -1;
+#ifdef TARGET_SUNOS
+    // On illumos/Solaris, the recommended way to obtain machine
+    // architecture is using `sysinfo` rather than `utsname.machine`.
+
+    char isa[32];
+    if (sysinfo(SI_ARCHITECTURE_K, isa, sizeof(isa)) > -1)
+    {
+#else
+    struct utsname _utsname;
+    if (uname(&_utsname) > -1)
+    {
+        char* isa = _utsname.machine;
+#endif
+        // aarch64 or arm64: arm64
+        if (strcmp("aarch64", isa) == 0 || strcmp("arm64", isa) == 0)
+        {
+            result = ARCH_ARM64;
+        }
+
+        // starts with "armv6" (armv6h or armv6l etc.): armv6
+        else if (strncmp("armv6", isa, strlen("armv6")) == 0)
+        {
+            result = ARCH_ARMV6;
+        }
+
+        // starts with "arm": arm
+        else if (strncmp("arm", isa, strlen("arm")) == 0)
+        {
+            result = ARCH_ARM;
+        }
+
+        // x86_64 or amd64: x64
+        else if (strcmp("x86_64", isa) == 0 || strcmp("amd64", isa) == 0)
+        {
+#ifdef TARGET_OSX
+            int is_translated_process = 0;
+            size_t size = sizeof(is_translated_process);
+            if (sysctlbyname("sysctl.proc_translated", &is_translated_process, &size, NULL, 0) == 0 && is_translated_process == 1)
+                result = ARCH_ARM64;
+            else
+#endif
+            result = ARCH_X64;
+        }
+
+        // ix86 (possible values are i286, i386, i486, i586 and i686): x86
+        else if (isa[0] == 'i' && isa[2] == '8' && isa[3] == '6')
```

Comment:
In theory, this can access uninitialized memory now.  The static analysis tools are likely to complain about it.

What was wrong with the `strlen` call that was here before?

Reply:
Ah, good catch. I thought it's a leftover from previous revision. 
```suggestion
        else if (strlen(isa) == strlen("i386") && isa[0] == 'i' && isa[2] == '8' && isa[3] == '6')
```